src/app/loadout-builder/process-worker/process.ts

```diff
@@ -345,27 +361,53 @@ export function process(
                 }
                 index++;
               }
+              energyRemaining[armorIndex] = item.energy
+                ? item.energy.capacity - item.energy.val
+                : 0;
+              armorIndex++;
             }
 
+            // Check stat tiers
             let totalTier = 0;
             let statRangeExceeded = false;
+            const statMods: number[] = [];
             for (const statHash of statOrder) {
-              const value = stats[statHash];
-              const tier = statTier(value);
-
-              // Update our global min/max for this stat
-              const range = statRanges[statHash];
-              if (value > range.max) {
-                range.max = value;
-              }
-              if (value < range.min) {
-                range.min = value;
-              }
+              let value = stats[statHash];
+              let tier = statTier(value);
 
               const filter = statFilters[statHash];
               if (!filter.ignored) {
-                if (tier > filter.max || tier < filter.min) {
+                if (tier > filter.max) {
                   statRangeExceeded = true;
+                } else {
+                  // Automatically add stat mods to taste
+                  if (generalModsAvailable - statMods.length > 0) {
+                    const newValue = addStatMods(
+                      statHash,
+                      generalModsAvailable,
+                      value,
+                      filter.max,
+                      energyRemaining,
+                      statMods
+                    );
+                    value = newValue;
+                    stats[statHash] = value;
+                  }
+
+                  tier = statTier(value);
+
+                  if (tier < filter.min) {
+                    statRangeExceeded = true;
+                  }
+                }
```

Comment:
I feel like this needs to pushed inside of `canTakeSlotIndependentMods`. The reason for this is we have no idea if an item is already using up all its energy. For example if I max out my energy in a class item, we get no stat mods.

<img width="923" alt="Screen Shot 2021-11-15 at 7 17 07 am" src="https://user-images.githubusercontent.com/7344652/141697209-5c040f2f-4243-483d-be7a-cb489212e656.png">

Reply:
Actually now that I have thought about this, I think it would be better to do this at render time rather than in process. Reason for this is if we do it here we lose the ability to exit early when we find any combination of mods that will fit (that is what `canTakeSlotIndependantMods` does). If we do it at render time we can make it a part of the search to find the best assignment of mods. I would do it in [this if loop](https://github.com/DestinyItemManager/DIM/blob/master/src/app/loadout/mod-utils.ts#L276-L284).

Reply:
I'm afraid I don't quite follow either of your comments. Can you spell out exactly what you're suggesting? Is it that we should consider the energy used by the mods chosen by the user before applying the automatic stat mods?

Reply:
As discussed outside this, the issue is a lot harder due to correct sorting of sets. Not sure I have a solution just yet, will report back once I have thought about it a bit.